### ISSUE TITLE
support for non-LaTeX greek symbols

### ISSUE THREAD

**Issue by pzinn** [Khan/KaTeX#1282]
Currently, greek letters can be input either as TeX (e.g., \alpha) or directly as unicode (\u03b1), but in both cases *only if they're defined in TeX*. Is there any reason to forbid greek letters that TeX didn't define, e.g., capital Eta? They could still be input as e.g. \u0397. The change to symbols.js would be minimal.

**Comment by pzinn** [comment]
Curiously, \omicron is actually defined in symbols.js even though it doesn't exist in TeX.
My original idea was rather to only allow the unicode version, but of course one can also introduce new TeX equivalent \xxx in the process.

**Comment by edemaine** [comment]
@pzinn Our primary goal is LaTeX compatibility -- but LaTeX consists of a *lot* of packages.  According to [this StackExchange post](https://tex.stackexchange.com/questions/233257/omicron-not-working-in-latex), `\omicron` is defined by `unicode-math` in XeLaTeX.  It looks like capital Eta (Unicode symbol and as `\mupEta`) is also defined there (see http://ftp.math.purdue.edu/mirrors/ctan.org/macros/latex/contrib/unicode-math/unimath-symbols.pdf#page=82) and we should do the same. Probably worth a review of the Greek letters in general to see if we're missing any others. @ronkok can comment more.

**Comment by ronkok** [comment]
Yes capital letter eta, code point U+0397, is supported by the `unicode-math` package, so it is a legitimate target for KaTeX.  So are the other Greek letters.

But the KaTeX fonts do not contain a glyph at code point U+0397. So instead of using `symbols.js`, it would be cleaner to write it as a macro, say:
```
defineMacro("\\Eta", "H");
defineMacro("\u0397", "H");
```

**Comment by pzinn** [comment]
I have several comments. First, greek caps are normally *not* italic, try \Gamma for example.
The second thing is, I tried \u0397 (again, not italic, math) and it looks just fine to me. Using "H" instead is a bit of a hack.

**Comment by ronkok** [comment]
@pzinn You're right about the Greek caps being non-italic, of course. Silly error on my part. Thank you for pointing that out. I'll fix it.

**Comment by pzinn** [comment]
in fact I might as well give you the whole list
    0x41 => 0x391,          # (missing Alpha)
    0x42 => 0x392,          # (missing Beta)
    [0,1] => 0x393,         # \Gamma, \Delta
    0x45 => 0x395,          # (missing Epsilon)
    0x5A => 0x396,          # (missing Zeta)
    0x48 => 0x397,          # (missing Eta)
    2 => 0x398,             # \Theta
    0x49 => 0x399,          # (missing Iota)
    0x4B => 0x39A,          # (missing Kappa)
    3 => 0x39B,             # \Lambda
    0x4D => 0x39C,          # (missing Mu)
    0x4E => 0x39D,          # (missing Nu)
    4 => 0x39E,             # \Xi
    0x4F => 0x39F,          # (missing Omicron)
    5 => 0x3A0,             # \Pi
    0x50 => 0x3A1,          # (missing Rho)
    6 => 0x3A3,             # \Sigma
    0x54 => 0x3A4,          # (missing Tau)
    [7,8] => 0x3A5,         # \Upsilon, \Phi
    0x58 => 0x3A7,          # (missing Chi)
    [9,0xA] => 0x3A8,       # \Psi, \Omega

**Comment by edemaine** [comment]
@pzinn Good alternative to raise.  One possible disadvantage of adding these symbols to our fonts would be that it would make the font files larger (unless fonts can have symlinks between symbols?). I also don't think the mappings you give will do what we want, because they will remove the existing definitions of Latin A, etc. (I'm pretty sure those are mappings, and thus do not allow duplication.) Of course, we could change the font code in more complicated ways, but we could also do that in the future if this becomes a big concern.

I'm curious what unicode-math/XeLaTeX does.  Are these defined in the font, or are they macro'd to Latin letters?  Ah, it turns out that they are macrod, in `engrec/engrec.sty`:
```latex
\providecommand{\upomicron}{\text{\rmfamily\upshape o}}
```
This seems like further justification for the current approach...

**Comment by pzinn** [comment]
have a look at omicron:
``
    [0x61,0x7A] => 0x61,    # a - z
    0x6F => 0x3BF,          # omicron
``
it definitely redefines o

**Comment by edemaine** [comment]
@pzinn I stand corrected! Hmm, if there's some prior art to this approach, then we should definitely consider this alternative approach...

**Comment by edemaine** [comment]
I did some experiments with `unicode-math` in XeLaTeX and how they interact with font commands:
```latex
$ο Ο$ mathnormal

$\mathit{ο Ο}$ mathit

$\mathrm{ο Ο}$ mathrm

$\mathbf{ο Ο}$ mathbf
```
![image](https://user-images.githubusercontent.com/2218736/39576623-c15a1a08-4eac-11e8-9735-84cc8cef7021.png)

So, as expected, upper-case Greek renders upright while lower-case Greek renders italic, while `\mathrm` makes them both upright.  But `\mathit` and `\mathbf` don't work!  Weird...

**Comment by ronkok** [comment]
Well in general, `\mathit` and `mathbf` don't stack the way that `\textit` and `\textbf` will. Those LaTeX font stacking nuances mean that the results of PR #1284 will not behave exactly like an actual glyphs in the KaTeX fonts. In that regard, @pzinn is correct. Adding glyphs to the KaTeX fonts would result in better behavior.

Which would mean a breaking change if PR #1284 is accepted now but a font glyph gets added later. So it would be entirely appropriate to deny PR #1284 and wait for someone to change the font. An aversion to breaking changes is what delayed Unicode characters in general for 18 months.

I don't do fonts. PR #1284 was my attempt to be helpful, but I realize that it has flaws and I'm not terribly attached to it. We have many outstanding issues more important than these particular characters.

**Comment by edemaine** [comment]
#1285 fixes the main issue here.  There remains a question of whether to duplicate more characters to their capital Greek equivalents, for copy/paste improvement (and anything else? MathML I guess?), matching `\omicron`.

**Comment by pzinn** [comment]
In case you do decide to go ahead with the font change, let me point out the necessary changes (tested on my fork), before I forget them:
- replace in makeFF (fonts) *and* mapping.pl (metrics) the partial capital greek list with the updated one mentioned above
- in symbols.js replace the latin capital letters of your latest PR with the unicode itself (that's the lazy way; I did it slightly differently but the end result is the same)

**Comment by ronkok** [comment]
I think this issue was addressed by #1285.
